### PR TITLE
Add support to register amis 

### PR DESCRIPTION
**Pull request by jimbydamonk**
This add support for the register amis from snapshots or s3. 
Supports all the setting except for dry run from http://boto.readthedocs.org/en/latest/ref/ec2.html

### CODE REVIEW

**Comment by gregdek** [comment]
@scicoin-project Please review this PR to make sure it adheres to the following guidelines:

http://docs.ansible.com/developing_modules.html#module-checklist

If it passes these guidelines, and if you believe it’s a good PR otherwise, please add a comment with "shipit" in the text, and we will flag it for inclusion.

If it fails these guidelines, please add a comment with "needs_revision" in the text, along with the needed changes, and we will flag it for revision.

**Comment by juliedavila** [comment]
@jimbydamonk can you rebase please?

**Comment by jimbydamonk** [comment]
Rebased. Let me know if there is something else I should do. Thanks!

**Comment by gregdek** [comment]
Thanks @jimbydamonk. To the current maintainers, @jjshoe please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit' or 'needs_revision' as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by jjshoe** [comment]
Sorry for the delay, reviewing this today, will be interesting as this is implemented in both #1786 and #2029

File: cloud/amazon/ec2_ami.py [hunk 1]
```diff
@@ -87,7 +87,68 @@
...
+    version_added: "2.0"
+  kernel_id:
+    description:
+      - The ID of the kernel with which to launch the instances
+    required: false
+    default: null
+    version_added: "2.0"
+  root_device_name:
+    description:
+      - The root device name (e.g. /dev/sda1)
```
Comment (jjshoe) [inline cloud/amazon/ec2_ami.py]: /dev/sda1 could be wrapped in C()

File: cloud/amazon/ec2_ami.py [hunk 2]
```diff
@@ -87,7 +87,68 @@
...
+    required: false
+    default: null
+    version_added: "2.0"
+  virtualization_type:
+    description:
+      - The virutalization_type of the image.
```
Comment (jjshoe) [inline cloud/amazon/ec2_ami.py]: virtualization_type should be wrapped in C()

File: cloud/amazon/ec2_ami.py [hunk 3]
```diff
@@ -87,7 +87,68 @@
...
+    choices: [ "hvm", "paravirtual" ]
+    required: true
+    default: hvm
+    version_added: "2.0"
+  sriov_net_support:
+    description:
+      - Advanced networking support.
```
Comment (jjshoe) [inline cloud/amazon/ec2_ami.py]: Be more descriptive, it's not clear what the default behavior is, my options are no advanced networking support, and simple advanced networking support?

File: cloud/amazon/ec2_ami.py [hunk 4]
```diff
@@ -87,7 +87,68 @@
...
+    choices: [ None, "simple" ]
+    required: false
+    default: null
+    version_added: "2.0"
+  snapshot_id: 
+    description:
+      - A snapshot ID for the snapshot to be used as root device for the image. Mutually exclusive with device_mapping, requires root_device_name
```
Comment (jjshoe) [inline cloud/amazon/ec2_ami.py:139]: options need to be wrapped in C()

Comment (jjshoe) [inline cloud/amazon/ec2_ami.py:139]: "The ID of a snapshot to use as the root device image." or something, current wording is difficult to parse.

File: cloud/amazon/ec2_ami.py [hunk 5]
```diff
@@ -87,7 +87,68 @@
...
+    required: false
+    default: null
+    version_added: "2.0"
+  delete_root_volume_on_termination:
+    description:
+      - Whether to delete the root volume of the image after instance termination. Only applies when creating image from snapshot_id. Defaults to False. Note that leaving volumes behind after instance termination is not free.
```
Comment (jjshoe) [inline cloud/amazon/ec2_ami.py]: make use of C()

File: cloud/amazon/ec2_ami.py [hunk 6]
```diff
@@ -150,6 +211,15 @@
           no_device: yes
   register: instance
 
+# Register AMI from snapshot
```
Comment (jjshoe) [inline cloud/amazon/ec2_ami.py]: How about an example using a manifest out of a S3 bucket as well?

**Comment by jjshoe** [comment]
Things I think need to be addressed:

1) All comments on the file
2) When I ran the example twice in a row in quick succession I got '''msg: Error in retrieving duplicate AMI details'''
3) When I run it twice in a row with a gap in between the second run takes forever to run, and returns "ok", not sure why so slow? It should just check if the ami already exists and bails?
4) Need to list module return values in doc
5) Provide a way to test manifest created amis

needs_revision

http://docs.ansible.com/ansible/developing_modules.html#module-checklist

**Comment by jjshoe** [comment]
A good chunk of the return structure will end being documented by this: #2320

**Comment by gregdek** [comment]
Thanks @jimbydamonk for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
Thanks @jimbydamonk for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
Thanks @jimbydamonk for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@jimbydamonk A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@jimbydamonk Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@jimbydamonk A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@jimbydamonk Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@jimbydamonk A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@jimbydamonk Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by jimbydamonk** [comment]
I have rebase and update the documentation as per @jjshoe's suggestions. 

I just ran a few tests and could not duplicate the  '''msg: Error in retrieving duplicate AMI details'''
or the slowness. I think the changes made since I have touch this have solved them. 

The return info is also up to date. 
ready_for_review

**Comment by gregdek** [comment]
Thanks @jimbydamonk. To the current maintainers, @jjshoe please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit', 'needs_revision' or 'close_me' as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

File: cloud/amazon/ec2_ami.py [hunk 7]
```diff
@@ -381,8 +467,13 @@ def create_image(module, ec2):
                 bd = BlockDeviceType(**device)
                 bdm[device_name] = bd
             params['block_device_mapping'] = bdm
+            register_params['block_device_mapping'] = bdm
+
+        if instance_id:
```
Comment (jjshoe) [inline cloud/amazon/ec2_ami.py:472]: Nit picky, should this be an if/else if?

**Comment by jjshoe** [comment]
shipit

File: cloud/amazon/ec2_ami.py (hunk 7 above)
Comment (jimbydamonk) [inline cloud/amazon/ec2_ami.py:472]: you are right. This check should really be checking to see what the action is (create or register) I will update it tonight.

**Comment by gregdek** [comment]
Thanks again to @jimbydamonk for this PR, and thanks @jjshoe for reviewing. Marking for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

File: cloud/amazon/ec2_ami.py [hunk 8]
```diff
@@ -544,14 +644,49 @@ def main():
...
+            # Changed is always set to true when provisioning new AMI
+            if not module.params.get('instance_id'):
+                module.fail_json(msg='instance_id is required for new image')
+
+        if module.params.get('action') == "register":
+            if not module.params.get('virtualization_type'):
+                module.fail_json(
+                    msg='virtualization_type is required for new image')
+
+            if module.params.get('virtualization_type') not in ['paravirtual', 'hvm']:
```
Comment (jimi-c) [inline cloud/amazon/ec2_ami.py:660]: Things like this should be set in the params above using the `choices` option to limit the selections.

**Comment by gregdek** [comment]
Thanks @jimbydamonk. To the current maintainers, @jjshoe please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit', 'needs_revision' or 'close_me' as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

**Pull request closed without merging**